typescript/src/digital-factory.ts

```diff
@@ -303,4 +303,88 @@ export class DigitalFactoryDemo {
         const response = await this.httpGetDigitalFactory(`${API_ROOT_URL}/cura/v1/projects?${query}`);
         return response.data;
     }
+
+    async generateReport(clusterIds: string[]): Promise<string> {
+        await this._checkTokenExpiration();
+
+        print('Requesting report generation');
+        const today = new Date();
+        const sevenDaysAgoMs = Date.now() - 24 * 60 * 60 * 7 * 1000;
+        const sevenDaysAgo = new Date(sevenDaysAgoMs);
+        const response = <{ data: ReportStatus }> await this.httpPutDigitalFactory(`${API_ROOT_URL}/report/v1/reports`,
+            {
+                data: {
+                    cluster_ids: clusterIds,
+                    start_date: sevenDaysAgo.toISOString(),
+                    end_date: today.toISOString(),
+                    report_type: 'print_jobs',
+                },
+            });
+
+        let status: ReportStatus = response.data;
+        const reportID = status.report_id;
+        while (status.status === 'pending' || status.status === 'processing') {
+            print('Waiting one second');
+            await new Promise((resolve) => setTimeout(resolve, 1000));
+
+            print('Checking the status of the report');
+            const statusResponse = <{ data: ReportStatus }> await this.httpGetDigitalFactory(`${API_ROOT_URL}/report/v1/reports/${reportID}`);
+            status = statusResponse.data;
+        }
```

Comment:
Await in a loop is usually bad practice... And the code is repeating a lot. I feel like the polling here should be extracted away from the main logic. e.g.
```suggestion
        const result = await poll(
            () => this.httpGetDigitalFactory(`${API_ROOT_URL}/report/v1/reports/${reportID}`),
            response => !['pending', 'processing'].includes(response.status)
        )
        const status: ReportStatus = result.status;
        const reportID = status.report_id;
```

Reply:
There is nothing wrong with an `await` in a loop. I don't know where that idea is coming from. You could say that this loop should give up after a certain amount of time though. But the purpose of the code is to be readable first and to teach what the happy flow looks like. Introducing a `poll()` function makes sense and leans DRY, but gets in the way of teaching. So I wouldn't do it.